src/coprocessor/endpoint.rs

```diff
@@ -428,69 +377,127 @@ impl Drop for RequestTracker {
     }
 }
 
-#[derive(Debug)]
 pub struct RequestTask {
-    req: Request,
-    cop_req: CopRequest,
-    ctx: ReqContext,
+    req_ctx: Arc<ReqContext>,
+    req_handler_builder: RequestHandlerBuilder,
     on_resp: OnResponse<Response>,
     tracker: RequestTracker,
 }
 
+impl Debug for RequestTask {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("RequestTask")
+            .field("req_ctx", &self.req_ctx)
+            .field("tracker", &self.tracker)
+            .finish()
+    }
+}
+
+impl Display for RequestTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
 impl RequestTask {
     pub fn new(
         peer: String,
-        req: Request,
+        mut req: Request,
         on_resp: OnResponse<Response>,
         recursion_limit: u32,
+        batch_row_limit: usize,
+        max_handle_duration: Duration,
     ) -> Result<RequestTask> {
-        let mut table_scan = false;
+        let (data, ranges, context, tp) = (
+            req.take_data(),
+            req.take_ranges(),
+            req.take_context(),
+            req.get_tp(),
+        );
+        // drop it in case of mistakenly using its fields
+        drop(req);
+
+        let mut is_table_scan = false;
         let mut is_desc_scan: Option<bool> = None; // only used in slow query logs
-        let (start_ts, cop_req) = match req.get_tp() {
+        let (req_ctx, req_handler_builder): (_, RequestHandlerBuilder) = match tp {
             REQ_TYPE_DAG => {
-                let mut is = CodedInputStream::from_bytes(req.get_data());
+                let mut is = CodedInputStream::from_bytes(&data);
                 is.set_recursion_limit(recursion_limit);
                 let mut dag = DAGRequest::new();
                 box_try!(dag.merge_from(&mut is));
                 if let Some(scan) = dag.get_executors().iter().next() {
                     // the first executor must be table scan or index scan.
-                    table_scan = scan.get_tp() == ExecType::TypeTableScan;
-                    if table_scan {
+                    is_table_scan = scan.get_tp() == ExecType::TypeTableScan;
+                    if is_table_scan {
                         is_desc_scan = Some(scan.get_tbl_scan().get_desc());
                     } else {
                         is_desc_scan = Some(scan.get_idx_scan().get_desc());
                     }
                 }
-                (dag.get_start_ts(), CopRequest::DAG(dag))
+                let req_ctx = Arc::new(ReqContext::new(
+                    context,
+                    dag.get_start_ts(),
+                    is_table_scan,
+                    max_handle_duration,
+                ));
+                let req_ctx_clone = Arc::clone(&req_ctx);
+                let ranges = ranges.to_vec();
```

Comment:
It's repeated in each match case too, could we move it before `match`?

Reply:
I guess not, we uses this `ranges` (the range before `to_vec()`) later when building the request tracker. Also due to the borrowing checker, we cannot leave them together.